Granger causality is a statistical hypothesis test used to determine whether one time series can predict another. It does not imply actual causality in the common sense but measures whether past values of one variable contain information that helps predict future values of another variable, beyond the information already contained in the past values of the variable being predicted.

In technical terms:

If a time series X "Granger-causes" another time series Y, it means that past values of X provide statistically significant information about future values of Y, improving the predictions of Y compared to when only past values of Y are used.
The test involves comparing two models: one that uses past values of both X and Y to predict Y, and another that uses only past values of Y to predict itself. If the first model significantly improves predictions, X is said to Granger-cause Y.
It is commonly used in econometrics and time series analysis, especially in fields like finance and economics, where determining the direction of relationships between variables can be important.

Key points about Granger causality:

It requires stationarity (constant mean and variance over time) of the time series involved.
It only detects temporal precedence—it cannot confirm true causality.
It is often applied to systems where feedback loops may exist, meaning causality could go both ways.
Granger causality tests are usually performed using statistical software, which outputs p-values indicating whether one series Granger-causes the other.

Sure! Let’s go through an example of how Granger causality can be tested in a simple time series context.

Example: Testing if Stock Prices Granger-Cause Economic Growth
Let’s say you have two time series:

Stock Prices (X) – Monthly average stock prices.
GDP Growth (Y) – Monthly GDP growth rates.
You want to test whether past values of Stock Prices can help predict future values of GDP Growth, i.e., whether Stock Prices Granger-cause GDP Growth.

Steps to perform Granger Causality Test:
Collect the Data: Obtain time series data for both stock prices and GDP growth over the same time period, say over 10 years, on a monthly basis.

Check Stationarity: Before testing, you need to ensure both time series are stationary (i.e., their statistical properties like mean and variance are constant over time). You can apply transformations or difference the data if it’s non-stationary.

Lag Selection: Choose how many past values (lags) of each variable to include in the model. For instance, you might decide to include the past 2 months' values for stock prices and GDP growth (lag of 2).

Set up two models:

Restricted Model (without stock prices): Only use past GDP growth values to predict future GDP growth.
Unrestricted Model (with stock prices): Use both past values of stock prices and past values of GDP growth to predict future GDP growth.
Run the Granger Causality Test:

If the unrestricted model (with stock prices) significantly improves the prediction of GDP growth (based on the F-test or p-values), then stock prices Granger-cause GDP growth.
If not, you fail to reject the null hypothesis, meaning stock prices don’t Granger-cause GDP growth.
Hypotheses:
Null Hypothesis: Stock prices do not Granger-cause GDP growth.
Alternative Hypothesis: Stock prices Granger-cause GDP growth.
Interpretation:
If the p-value is less than your significance level (say, 0.05), you reject the null hypothesis and conclude that stock prices Granger-cause GDP growth.
If the p-value is greater than 0.05, you fail to reject the null hypothesis, implying there is no evidence that stock prices help predict GDP growth.

Example Output (Hypothetical):

Test Statistic	p-value	Conclusion

4.56	0.03	Reject Null Hypothesis

In this case, the p-value is 0.03, which is less than 0.05. Thus, you would reject the null hypothesis and conclude that stock prices Granger-cause GDP growth.

This is just an illustrative example. In practice, you would use statistical software (like Python’s statsmodels library) to perform the Granger causality test with actual data.

In [1]:
import numpy as np
import pandas as pd
from statsmodels.tsa.stattools import grangercausalitytests

# Example Data
# Let's generate some sample data for Stock Prices (X) and GDP Growth (Y)
np.random.seed(0)
data_length = 100
time = np.arange(data_length)

# Simulate Stock Prices (X)
X = np.sin(0.1 * time) + np.random.normal(0, 0.1, data_length)

# Simulate GDP Growth (Y) with some dependence on past values of X
Y = 0.5 * np.roll(X, 1) + np.random.normal(0, 0.1, data_length)

# Create a DataFrame
data = pd.DataFrame({'Stock_Prices': X, 'GDP_Growth': Y})

# Granger Causality Test
# We test if past values of Stock Prices (X) Granger-cause GDP Growth (Y)
max_lag = 2  # We will test for 2 lags (you can adjust this)
test_result = grangercausalitytests(data[['GDP_Growth', 'Stock_Prices']], max_lag, verbose=True)

# Output the result
print(test_result)


Granger Causality
number of lags (no zero) 1
ssr based F test:         F=142.7494, p=0.0000  , df_denom=96, df_num=1
ssr based chi2 test:   chi2=147.2103, p=0.0000  , df=1
likelihood ratio test: chi2=90.1956 , p=0.0000  , df=1
parameter F test:         F=142.7494, p=0.0000  , df_denom=96, df_num=1

Granger Causality
number of lags (no zero) 2
ssr based F test:         F=49.1575 , p=0.0000  , df_denom=93, df_num=2
ssr based chi2 test:   chi2=103.6008, p=0.0000  , df=2
likelihood ratio test: chi2=70.6895 , p=0.0000  , df=2
parameter F test:         F=49.1575 , p=0.0000  , df_denom=93, df_num=2
{1: ({'ssr_ftest': (142.7493999364213, 1.0622210540064497e-20, 96.0, 1), 'ssr_chi2test': (147.21031868443447, 7.059187685413867e-34, 1), 'lrtest': (90.19556117891958, 2.1574556585141538e-21, 1), 'params_ftest': (142.74939993642113, 1.0622210540064853e-20, 96.0, 1.0)}, [<statsmodels.regression.linear_model.RegressionResultsWrapper object at 0x7714061ff350>, <statsmodels.regression.linear_model.Regr

/home/yosakoi/miniconda3/envs/tradingbot/lib/python3.12/site-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print results
  warnings.warn(


his output indicates the results of a Granger causality test with two lags, showing whether Stock Prices Granger-cause GDP Growth based on different test statistics.

Let's break it down:

Key Statistics in the Output:
SSR Based F-Test:

F-statistic and p-value: These values come from an F-test that compares the restricted and unrestricted models. If the p-value is very small (e.g., close to 0), it suggests that the inclusion of past values of stock prices significantly improves the model predicting GDP growth.
For lag 1: F = 142.7494, p-value = 0.0000 (indicating strong significance).
For lag 2: F = 49.1575, p-value = 0.0000 (also strongly significant).
SSR Based Chi-Square Test:

This test is similar to the F-test but is based on the chi-square distribution. It checks the same hypothesis as the F-test but uses a different approach. A very small p-value also suggests significance.
For lag 1: chi-square = 147.2103, p-value = 0.0000.
For lag 2: chi-square = 103.6008, p-value = 0.0000.
Likelihood Ratio Test (LR Test):

This test compares the likelihoods of the restricted and unrestricted models. Again, a small p-value indicates significance.
For lag 1: chi-square = 90.1956, p-value = 0.0000.
For lag 2: chi-square = 70.6895, p-value = 0.0000.
Parameter F-Test:

This is another test on the coefficients of the model. If the p-value is small, it means that the parameters (in this case, the lags of stock prices) are significant predictors of GDP growth.
For lag 1: F = 142.7494, p-value = 0.0000.
For lag 2: F = 49.1575, p-value = 0.0000.
Interpretation of the Results:
For both lags 1 and 2, all tests (F-test, chi-square test, likelihood ratio test) have p-values of 0.0000 (or very close to zero), meaning that past values of stock prices significantly improve the prediction of GDP growth.
You can reject the null hypothesis (which states that stock prices do not Granger-cause GDP growth) for both lag 1 and lag 2 models.
What It Means:
Stock Prices Granger-cause GDP Growth: The test strongly indicates that stock prices can be used to predict GDP growth. The low p-values across the different tests show that past stock prices contain valuable information that can predict future GDP growth.

Lag Considerations: The test results for both 1 lag and 2 lags show significance, so you might want to consider both lag structures in your final model. You could choose the number of lags based on additional model fit criteria (like AIC or BIC), but both lag models suggest a strong relationship.

In short, based on this output, Stock Prices do Granger-cause GDP Growth, and you can confidently reject the null hypothesis across different lag structures.